python/arb.cpp

```diff
@@ -0,0 +1,235 @@
+        .def("__str__",  &node_info_string)
+        .def("__repr__", &node_info_string);
+
+    // get_node_info
+    m.def("get_node_info", &arb::hw::get_node_info,
+        "Returns a description of the hardware resources available on the host compute node.");
+
+    //
+    // models
+    //
+    pb::class_<arb::model> model(m, "model", "An Arbor model.");
+
+    model
+        // A custom constructor that wraps a python recipe with
+        // arb::py_recipe_shim before forwarding it to the arb::recipe constructor.
+        .def(pb::init(
+                [](std::shared_ptr<arb::py::recipe>& r, const arb::domain_decomposition& d) {
+                    return new arb::model(arb::py_recipe_shim(r), d);
+                }),
+                // Release the python gil, so that callbacks into the python
+                // recipe r don't deadlock.
+                pb::call_guard<pb::gil_scoped_release>(),
+                "Initialize the model described by a recipe, with cells and network "
+                "distributed according to decomp.",
+                "recipe"_a, "decomp"_a)
+        .def("reset", &arb::model::reset,
+                pb::call_guard<pb::gil_scoped_release>(),
+                "Reset the model to its initial state to rerun the simulation again.")
+        .def("run", &arb::model::run,
+                pb::call_guard<pb::gil_scoped_release>(),
+                "Advance the model state to time tfinal, in time steps of size dt.",
+                "tfinal"_a, "dt"_a);
+
+    //
+    // metering
+    //
+    pb::class_<arb::util::measurement> measurement(m, "measurement",
+             "Describes the recording of a single statistic over the course of a simulation,\n"
+             "gathered by the meter_manager.");
+    measurement.def_readwrite("name", &arb::util::measurement::name,
+                    "Descriptive label of the measurement, e.g. 'wall time' or 'memory'.")
+               .def_readwrite("units", &arb::util::measurement::units,
+                    "SI units of the measurement, e.g. s or MiB.")
+               .def_readwrite("measurements", &arb::util::measurement::measurements,
+                    "A list of measurements, with one entry for each checkpoint.\n"
+                    "Each entry is a list of values, with one value for each domain (MPI rank).");
+
+    pb::class_<arb::util::meter_manager> meter_manager(m, "meter_manager");
+    meter_manager.def(pb::init<>())
+                     .def("start", &arb::util::meter_manager::start)
+                     .def("checkpoint", &arb::util::meter_manager::checkpoint);
+
+    // wrap meter_report type such that print(meter_report) works
+    pb::class_<arb::util::meter_report> meter_report(m, "meter_report");
+    meter_report.def("__str__", &meter_report_string)
+                .def("__repr__",&meter_report_string);
+
+    m.def("make_meter_report", &arb::util::make_meter_report,
+          "Generate a meter_report from a set of meters.");
+}
```

Comment:
The `measurement` and `meter_report` classes probably do not need to be exposed; a shim for the `meter_manager` that reports in JSON (using Python's JSON support) would be simpler.